### PR TITLE
MAID-1864: Merge the routing and vault event handling threads.

### PR DESCRIPTION
**Pull request by michaelsproul**
Issues to resolve:

* ~2 failing tests: `node_restart` and `failing_connections_group_of_three`.~
* ~Probably some formatting + linting.~

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @afck (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

**Comment by michaelsproul** [comment]
Rebasing now..

File: src/state_machine.rs [hunk 1]
```diff
@@ -245,35 +266,33 @@ impl Debug for StateMachine {
     }
 }
 
-#[cfg(not(feature = "use-mock-crust"))]
 impl StateMachine {
-    /// Run the event loop for sending and receiving messages. Blocks until
-    /// the core is terminated, so it must be called in a separate thread.
-    pub fn run(&mut self) {
-        while self.is_running {
+    #[cfg(not(feature = "use-mock-crust"))]
```
Comment (fizyk20) [inline src/state_machine.rs]: Why limit this method to the production code? The only difference to `try_step` I see is that it will block the thread - I'm not sure if that will ever be useful in tests, but I can't see any reason to ban this altogether.

File: tests/mock_crust/drop.rs [hunk 2]
```diff
@@ -30,20 +31,21 @@ fn drop_node(nodes: &mut Vec<TestNode>, index: usize) {
...
+                Err(_) => panic!("Event::NodeLost({:?}) not received", name),
             }
         }
     }
 }
 
 #[test]
 fn failing_connections_group_of_three() {
-    let min_group_size = 8;
+    let _ = maidsafe_utilities::log::init(false);
```
Comment (fizyk20) [inline tests/mock_crust/drop.rs]: I don't think we should leave log initializers in the code.

File: tests/mock_crust/drop.rs [hunk 3]
```diff
@@ -76,7 +78,8 @@ fn node_drops() {
 #[test]
 #[cfg_attr(feature = "clippy", allow(needless_range_loop))]
 fn node_restart() {
-    let min_group_size = 8;
+    let _ = maidsafe_utilities::log::init(false);
```
Comment (fizyk20) [inline tests/mock_crust/drop.rs]: Same as above.

File: tests/mock_crust/merge.rs [hunk 4]
```diff
@@ -15,13 +15,15 @@
...
+use maidsafe_utilities;
 use rand::Rng;
-use routing::Event;
+use routing::{Event, EventStream};
 use routing::mock_crust::Network;
 use super::{create_connected_nodes_until_split, poll_and_resend, verify_invariant_for_all_nodes};
 
 // See docs for `create_connected_nodes_with_cache_until_split` for details on `prefix_lengths`.
 fn merge(prefix_lengths: Vec<usize>) {
+    let _ = maidsafe_utilities::log::init(false);
```
Comment (fizyk20) [inline tests/mock_crust/merge.rs]: Same as above.

File: src/node.rs [hunk 5]
```diff
@@ -139,13 +114,13 @@ impl NodeBuilder {
                         this is not supported",
                        full_id.public_id().name());
 
-                let _ = event_sender.send(Event::Terminate);
+                // FIXME(michael): send this event?
```
Comment (fizyk20) [inline src/node.rs]: I think we should make a decision about this before merging.

File: src/event_stream.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    /// After calling poll, any events produced will be accessible via try_next_ev.
+    fn poll(&mut self) -> bool;
+}
+
+/// Trait for state machines and other event producers who produce multiple events at once.
+pub trait EventStepper {
+    type Item;
+
+    /// Produce multiple events and return them, or Err(()) if something goes wrong.
+    /// TODO: propagate actual errors better.
```
Comment (fizyk20) [inline src/event_stream.rs]: This might be a good idea ;) Looks like a big task, though, so maybe it will be better to leave it for another PR.

File: src/state_machine.rs (hunk 1 above)
Comment (michaelsproul) [inline src/state_machine.rs]: It's unreachable (and considered unused) when compiling with mock crust at the moment. We could always remove the `cfg` if it turns out to be useful in a test?

File: src/node.rs (hunk 5 above)
Comment (michaelsproul) [inline src/node.rs]: Yeah

**Comment by michaelsproul** [comment]
Ok, I've rebased and run `rustfmt`. Should be ready to merge once we decide what to do about the possibility of send failing in the client's thread.

File: src/node.rs (hunk 5 above)
Comment (michaelsproul) [inline src/node.rs]: I think we have to return an `Evented<RoutingActionSender, StateMachine>` from `StateMachine::new`, and update the calls to this function in `Node` and `Client` so that they push them on their internal buffer (or send them on the `event_sender` in the case of the non-test Client).

File: src/evented.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+/// Blanket impl of `ToEvented` for every sized type.
+impl<T, Ev: Debug> ToEvented<Ev> for T where T: Sized {}
+
+/// Explode if an attempt is made to drop events.
+impl<T, Ev: Debug> Drop for Evented<T, Ev> {
+    fn drop(&mut self) {
+        if !self.events.is_empty() {
+            error!("Events were dropped: {:?}", self.events);
+            panic!();
```
Comment (michaelsproul) [inline src/evented.rs]: The idea behind this panic is that dropping events should be considered a fatal bug... however, panicking in a `Drop` implementation will likely abort (as per the [docs](https://doc.rust-lang.org/std/ops/trait.Drop.html#panics)), and indeed in practice I found it would generate a `SIGILL` on my Mac.

An alternative would be to just print the error and not panic at all, but I worry this would make it too easy to miss an event drop bug?

File: src/lib.rs [hunk 8]
```diff
@@ -54,7 +54,7 @@
 //! to the network through any node, and exchange public keys with it. That node becomes a
 //! bootstrap node for the client, and messages to and from the client will be routed over it.
 //!
-//! ```no_run
+//! ```ignore
```
Comment (michaelsproul) [inline src/lib.rs]: It can in non-testing mode, but in testing mode, the constructor for `Client` no longer takes an event-sender argument. We could conditionally include different examples here using `#cfg[...]` and `#[doc = ...]` but I was hesitant to spend time doing that unless you think it's worthwhile.

File: src/node.rs [hunk 9]
```diff
@@ -416,124 +388,149 @@ impl Node {
 
     /// Returns the first `count` names of the nodes in the routing table which are closest
     /// to the given one.
-    pub fn close_group(&self,
+    pub fn close_group(&mut self,
```
Comment (michaelsproul) [inline src/node.rs]: Done!

File: src/state_machine.rs [hunk 10]
```diff
@@ -18,21 +18,23 @@
 use action::Action;
 use crust::{CrustEventSender, PeerId, Service};
 use crust::Event as CrustEvent;
+use event::Event;
+use evented::*;
```
Comment (michaelsproul) [inline src/state_machine.rs]: All gone!

File: src/evented.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    /// Extract the events to an event buffer.
+    pub fn extract_to_buf(mut self, buffer: &mut VecDeque<Ev>) -> T {
+        buffer.extend(self.take_events());
+        self.take_value()
+    }
+
+    /// Consume the Evented<T> and yield just its events.
+    ///
+    /// This is one of two methods that allows us to drop events (the other is extract_to_buf).
+    pub fn get_events(mut self) -> Vec<Ev> {
```
Comment (michaelsproul) [inline src/evented.rs]: Yeah, I like that better. Changed.

File: src/evented.rs (hunk 7 above)
Comment (michaelsproul) [inline src/evented.rs]: I've removed the panic entirely, as I doubt the abort will ever be helpful.

File: src/lib.rs (hunk 8 above)
Comment (afck) [inline src/lib.rs]: Ah, right! Not sure what's the best solution, but at some point it would be good to make this `no-run` again somehow, to make sure it stays up to date with the `Client` API.
